index.bs

```diff
@@ -78,4 +260,327 @@ interface StorageBucket {
 };
 </xmp>
 
+A {{StorageBucket}} has an associated [=/storage bucket=].
+
+<h3 id="storage-bucket-persistence">Persistence</h3>
+
+Issue: Merge with [[Storage#buckets]] which already defines [=bucket mode=].
+
+<div algorithm>
+
+The <dfn method for="StorageBucket">persist()</dfn> method steps are:
+
+1. Let |p| be [=a new promise=].
+
+1. Run the following steps [=in parallel=]:
+
+    1. Let |bucket| be [=this=]'s [=/storage bucket=].
+
+    1. If |bucket| has been removed, [=reject=] |p| with {{InvalidStateError}}.
+    
+    1. If |bucket|'s [=bucket mode=] is `"persistent"`, [=/resolve=] |p| with true.
+
+    1. Otherwise,
+
+        1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+        1. Let |permission| be the result of [=getting the current permission state=] with `"persistent-storage"` and |environment|.
+
+        1. If |permission| is "{{PermissionState/granted}}", then set |bucket|'s [=bucket mode=] to `"persistent"` and [=/resolve=] |p| with true.
+
+        1. Otherwise, [=/resolve=] |p| with false.
+
+1. Return |p|.
+
+</div>
+
+<div algorithm>
+
+The <dfn method for="StorageBucket">persisted()</dfn> method steps are:
+
+1. Let |p| be [=a new promise=].
+
+1. Run the following steps [=in parallel=]:
+
+    1. Let |bucket| be [=this=]'s [=/storage bucket=].
+
+    1. If |bucket| has been removed, [=reject=] |p| with an {{InvalidStateError}}.
+
+    1. Otherwise,
+
+        1. If |bucket|'s [=bucket mode=] is `"persistent"`, then [=/resolve=] |p| with true.
+
+        1. Otherwise, [=/resolve=] |p| with false.
+
+1. Return |p|.
+
+</div>
+
+<h3 id="storage-bucket-quota">Quota</h3>
+
+A [=/storage bucket=] has a <dfn for=StorageBucket>quota value</dfn>, a number-or-null, initially null.
+Specifies the upper limit of usage in bytes which can be used by the bucket. The user agent MAY further
+limit the realized storage space.
+
+The <dfn>storage usage</dfn> of a [=/storage bucket=] is an [=implementation-defined=] rough estimate
+of the number of bytes used by all of its [=/storage bottle=]s.
+
+<div algorithm>
+
+The <dfn method for="StorageBucket">estimate()</dfn> method steps are:
+
+1. Let |p| be [=a new promise=].
+
+1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+1. Let |shelf| be the result of running [=obtain a local storage shelf=] with |environment|.
+
+1. If |shelf| is failure, [=reject=] p with an {{UnknownError}}.
+
+1. Otherwise, run the following steps [=in parallel=]:
+
+    1. Let |bucket| be [=this=]'s [=/storage bucket=].
+
+    1. If |bucket| has been removed, [=reject=] |p| with an {{InvalidStateError}}.
```

Comment:
As has been noted elsewhere we need to define the removal concept.  I think we can just say there is an associated removal flag on the bucket object.  You can check it here and set it in a removal algorithm you write.

Reply:
Added!